### DISCUSSION TITLE
[API Request] Endpoint to retrieve metadata about multiple songs at once

### DISCUSSION THREAD

**Discussion by dweymouth** [opensubsonic/open-subsonic-api#6] [API requests]
### Type of change

API Extension

### Proposal description

I am proposing a new endpoint, `GetSongs` which should behave exactly like the existing `GetSong`, except that it accepts multiple IDs and returns info for all the songs requested.

This is useful for clients to refresh metadata on an arbitrary set of songs at once, e.g. for a local play queue. 

### Backward compatibility impact

New API, no backward compatability impact

### Backward compatibility

- [X] No backward compatibility impact.

### API details

The response payload should be the same as `GetSong` except with multiple nested song elements instead of just one.

### Alternative solutions

Other suggested name could be `GetSongsInfo`. Or an alternate solution is to just extend the existing `GetSong` endpoint to accept multiple IDs, and keep the naming even though it could now be used to get multiple songs.

**Comment by Tolriq** [comment]
Do not forget to select the proper type of change.

Not sure about the naming, this is one of the global missing feature that for example force me to use `search3` to get all the songs. Would keep that for that purpose if we one day decide on a proper paging system for those.

Maybe something more like `getSongsInfo`, `getAlbumsInfo` and `getArtistsInfo` (so with an s) that accept multiple ids? For consistency and more distinct naming?

With that said, I like the idea, the less request the more efficient things are.

But we should maybe also first define that api extension supporting servers must support post queries, to allow sending larger list of ids?

**Reply by dweymouth** [reply]
> But we should maybe also first define that api extension supporting servers must support post queries, to allow sending larger list of ids?

That could be a discussion all on its own. If we're doing that then the Create/UpdatePlaylist endpoints should support POST requests as well.

My use case for `GetSongs` wasn't to fetch the whole library but to fetch updated metadata on the tracks in the local play queue. I do think an additional endpoint to get all songs, with pagination, is needed, and I'd suggest the naming `GetSongList/GetSongList2`, for consistency with the GetAlbumList endpoints.

**Reply by Tolriq** [reply]
Well `getAlbumList` is an aberration and the exception here, there's `getGenres`, `getArtists`, `getPodcats`, `getMusicFolders`, ....

I understand the need and it's to get info for a list of things, this can be used also for albums and artists, it's requesting info, the naming I propose represent the need in a clear way. `getSongs` having a multipel ids filter does not match anything in current API, hence the proposal a new proper name scheme for the getXXInfo supporting multiple ids.

And yes the `post` suggestion was for the whole api. It seems most server do already support that it's just not documented yet.

**Reply by dweymouth** [reply]
Good point on the naming, yes `GetSongsInfo` could work, though unlike GetArtistInfo/AlbumInfo it wouldn't return additional info from last.fm, it would be the same payload as `GetSong` except supporting multiple IDs

**Reply by dweymouth** [reply]
Or, even keeping the same endpoint name, but just extending it to accept multiple ID parameters.

**Reply by Tolriq** [reply]
>  it wouldn't return additional info from last.fm,

For now, but it could be expanded to return last.fm or musicbrainzId data too if wanted in the future without impacts. 

> same endpoint name

From an API perspective this is bad having a single endpoint that can return either a single item or a list of items, it's not nice to deal with in many cases from the parsers point of view. (And impossible to properly describe by a schema, so IMO should be forbidden as we try to improve things)

**Reply by deluan** [reply]
> That could be a discussion all on its own. If we're doing that then the Create/UpdatePlaylist endpoints should support POST requests as well.

I think this is a point that needs clarification in the docs: The original Subsonic server (and all its forks, Airsonic, Madsonic...) accepts POST for *all* endpoints, and at least one client uses it: play:Sub. I also implemented this in Navidrome to support play:Sub

**Reply by Tolriq** [reply]
Don't know how the other servers support is, and the format of the post. Might worth a proper API definition.

**Reply by deluan** [reply]
We can basically document how it works in Subsonic.

**Reply by Tolriq** [reply]
Well if you know how be my guest :) Easy to add in api-reference.

**Reply by spl0k** [reply]
Supysonic also supports POST and this support was added something like 10 years ago because some clients used it. I don't remember which one, maybe it was the official Android app.

The format is basically each arg in the query string is now a form field.

**Comment by dweymouth** [comment]
I actually don't need this API anymore at the moment since I figured out how to properly update state in my local play queue model. ;) But it still would be a nice API for any clients that save a play queue, or local playlists, across restarts, and need to fetch updated play counts, ratings, etc. for those tracks.
